### ISSUE TITLE
[promise-function-async] False positives with functions that return `any`

### ISSUE THREAD

**Issue by sindresorhus** [typescript-eslint/typescript-eslint#369]
**Repro**

```JSON
{
  "rules": {
    "typescript/promise-function-async": "error"
  }
}
```

```TS
function trim(query) {
	return query.trim();
}
```

**Expected Result**

I expected the rule to only report on functions that return a promise.

**Actual Result**

It reports on functions returning `any`.

**Additional Info**

**Versions**

| package                            | version |
| ---------------------------------- | ------- |
| `@typescript-eslint/eslint-plugin` | `1.5.0` |
| `@typescript-eslint/parser`        | `1.5.0` |
| `TypeScript`                       | `3.3.3333` |
| `ESLint`                           | `5.15.3` |
| `node`                             | `10.15.1` |
| `npm`                              | `6.9.0` |

**Comment by j-f1** [comment]
This is technically correct since an object of type `any` could be a Promise. Perhaps there should be an option `allowAny`, on by default, that allows any to be used here?

**Comment by sindresorhus** [comment]
While it might be technically correct, I don't see how it's ever wanted behavior for this rule. It would mean you would have to add `async` to anything that returns `any`, even originally synchronous functions.

**Comment by ThomasdenH** [comment]
Does the lint trigger on `function a(): Promise<string> | string`? I think the rule should only trigger when it definitely only returns promises, not when it could return a promise.

**Comment by j-f1** [comment]
That’s why i suggested adding an option that would allow `any` by default but let the user disallow it. However, if it’s very unlikely to be a useful option, there’s no need to add it.

**Comment by G-Rath** [comment]
I've got the same problem with `axios` & class getters:

```
class Mine {
  public get t() {
    return Axios;
  }
}
```

This gets flagged as `should be marked with async` (which is actually not allowed with `get`).

**Comment by scottohara** [comment]
Also seeing this in a `webpack.config.js` that has functions that return various configurations of webpack plugins, e.g.

(note: this is from a project that has a mixed codebase of JS and TS files, and the webpack config is Javascript)

```javascript
function extractCss(hashFilename) {
	return new MiniCssExtractPlugin({filename: hashFilename ? "[name]-[chunkhash:6].css" : "[name].css"});
}

function defineAppConfig(maxDataAgeDays) {
	return new webpack.DefinePlugin({MAX_DATA_AGE_DAYS: maxDataAgeDays});
}
```

Both of the above functions trigger `@typescript-eslint/promise-function-async`, but as far as I can tell neither of these plugin constructors return `any` or `Promise` types?

**Comment by radical-edo** [comment]
I've hit this problem as well. For some reason the following is treated as promise

```ts
class Validator {
  private _joi = Joi.extend(joiPhoneNumber);
  public getJoi(): JoiObject {
    return this._joi;
  }
}
```

Linter has issue with the `getJoi()` method. Thinks it's a `async` function and wants for me to add the `async` keyword :/

**Comment by burtek** [comment]
I'm getting similar error on `v1.13.0`

```ts
import { Reducer } from 'redux';
const reducer: Reducer<Record<string, string>> = (state = {}, _action) => state;
```

![image](https://user-images.githubusercontent.com/2940958/61980219-5981b880-aff6-11e9-8974-c577be66b797.png)

**Versions**

| package                            | version |
| ---------------------------------- | ------- |
| `@typescript-eslint/eslint-plugin` | `1.13.0` |
| `@typescript-eslint/parser`        | `1.13.0` |
| `TypeScript`                       | `3.5.3` |
| `ESLint`                           | `6.1.0` |
| `node`                             | `10.14.2` |
| `npm`                              | `6.4.1` |

**Comment by bradzacher** [comment]
have you tried the `allowAny: false` option?
Also if you try the 2.0 canary version, it should have the above option as the default value.

**Comment by bradzacher** [comment]
(closing as this didn't get closed when #553 was merged)
